### DISCUSSION TITLE
Replacing Content

### DISCUSSION THREAD

**Discussion by amp-insider** [svrnm/DemoMonkey#234] [Q&A]
Hey there!

Firstly, wow what an amazing tool! Seriously changing the game for teams and people like myself.

There's a piece I'm curious about in DemoMonkey and changing content on the page. After attempting a variety of methods:
- replace
- replace image
- replaceAttribute
- !style

I haven't been able to strike success quite yet!

Here is a two minute video describing the specific situation in further detail: [Demo Monkey & Insider Video](https://drive.google.com/file/d/1SWBFycACCKDqiiapkjaVfA9YpYZ7ecEB/view?usp=drive_link)

Curious if there's anything that sticks out as a potential solution.

Thanks!

**Comment by svrnm** [comment]
Hey @amp-insider, 

thanks for your feedback! Let's talk about the image issue here, and figure out if there is a capability in DemoMonkey already to get what you want to accomplish, for the other issue with the right-click editor I would suggest to raise an issue as this is a bug for sure.

So, let's go through the list of options:
* the "normal" replace function only works on text, so this is not an option
* "replaceImage" is able to work on background images, it does detection on the value of the current image, so you could do something like `!replaceImage(https://domain.tld/path/to/your/current/image.png) = https://domain2.tld/path/to/new/image`, it also accepts wildcards, so you can play around a little bit with putting `*` in the left hand side, e.g. start with `!replaceImage(https://domain.tld/*) = ...` and see if it works at all
* "replaceAttribute" is not applicable in this context, since it looks for a text element and then manipulates the attributes of the enclosing html element, but since there is no text within that element you will not be able to use it
* Same for "style"
* An option that remains is [!querySelector](https://github.com/svrnm/DemoMonkey/blob/main/USAGE.md#replace-based-on-css-selector), which is what you tried with !style, it takes a css query selector, and an attribute and then updates it accordingly, e.g. `!querySelector([data-v-0ed1b90b], style.backgroundImage) = value`

I hope that helps!

**Reply by amp-insider** [reply]
@svrnm What a reply! Thank you for the detailed information.

I appreciate the specific on why the "normal" replace isn't functional here.

After reading your message the steps I took are:

- replaceImage (unsuccessful)

`!replaceImage(https://poshstreet.shop/cdn/shop/products/2015-04-03*) = https://images.pexels.com/photos/90946/pexels-photo-90946.jpeg`

- querySelector (unsuccessful)

`!querySelector([data-v-0ed1b90b], style.background-image) = https://images.pexels.com/photos/90946/pexels-photo-90946.jpeg`

I'm beginning to think there's an odd way in which these images are rendered on the page, preventing DemoMonkey from accessing them.

**Reply by svrnm** [reply]
That's indeed possible, we have an open issue for [shadow DOM](https://github.com/svrnm/DemoMonkey/issues/146) and there are also some cases with iframes or other embeddings where things can get funky. 

In the inspector view scroll up and check how the code is embedded, is it in the main frame or is there some surroundings like a shadow DOM or iframes? 

As a step to investigate further, what you also could do is create a super simple HTML file run it locally and see if the changes are applied to it:

```html
<html>
<body>
<div data-v-0ed1b19b class="w-1 h-1 product-image” style=“background -image: url(...)”></div>
</body>
</html>
```

Question: is this ID `data-v-0ed1b19b` static or will it be different every time you load the page?

It also seems that all(?) parent elements of this div have the same attribute set as well, maybe that's leading to some strange behavior with demo monkey?

**Reply by amp-insider** [reply]
Thanks again for taking a dive into this! Seems to be no iframes. Checking shadow DOM, however, is beyond the normal technical boundaries of my expertise. I'm unable to tell if these elements are or are not part of the shadow DOM.

The IDs are static. `data-v-0ed1b19b` is consistent each time the page is loaded. Although, an interesting piece noted here, which you've pointed out... all of the images and their parent elements use the same ID of `data-v-0ed1b19b`. 

I've consulted our team internally to understand how we can target these elements using the DemoMonkey plugin. There seems to be an odd/interesting implementation from our side on how these elements render.

Again, appreciate your assistance to this point, @svrnm! When/if we find an answer I'll report back here :)

**Reply by svrnm** [reply]
> Checking shadow DOM, however, is beyond the normal technical boundaries of my expertise. I'm unable to tell if these elements are or are not part of the shadow DOM.

If I remember correctly in the developer view it would say "# shadow-dom" or something like that, but don't very I don't think it is that.

One more thought, maybe the `[...]` makes demo monkey trouble since the config is a glorified ini file that uses them for sections, maybe also try 

```
!querySelector("[data-v-0ed1b90b]", style.background-image) = https://images.pexels.com/photos/90946/pexels-photo-90946.jpeg
```

Note, if you can share with me an URL where I can take a look, I can give it a try as well. If you prefer a private channel, you can send me a message via LinkedIn: https://www.linkedin.com/in/severinneumann/

**Reply by amp-insider** [reply]
Watched [this video](https://www.youtube.com/watch?v=sVBgpz1z9Ts) looking to see if that translates to these pages for shadow DOM elements, and it doesn't seem so. 

I am flattered you're open to checking out the page yourself, I'll drop you a note on LinkedIn as the page it gated behind logins. Thanks and happy holidays!
